src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -306,104 +357,526 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::flush_last_part()
 {
-  int delDataReturn = filter->get_d4n_cache()->deleteData(obj->get_key().get_oid()); 
+  last_part = true;
+  return handle_data(bl_rem, 0, bl_rem.length());
+}
 
-  if (delDataReturn < 0) {
-    ldpp_dout(save_dpp, 20) << "D4N Filter: Cache delete data operation failed." << dendl;
-  } else {
-    ldpp_dout(save_dpp, 20) << "D4N Filter: Cache delete data operation succeeded." << dendl;
+int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  auto rgw_get_obj_max_req_size = g_conf()->rgw_get_obj_max_req_size;
+
+  if (!last_part && bl.length() <= rgw_get_obj_max_req_size) {
+    auto r = client_cb->handle_data(bl, bl_ofs, bl_len);
+
+    if (r < 0) {
+      return r;
+    }
   }
 
+  //Accumulating data from backend store into rgw_get_obj_max_req_size sized chunks and then writing to cache
+  if (write_to_cache) {
+    const std::lock_guard l(d4n_get_data_lock);
+    rgw::d4n::CacheBlock block, existing_block;
+    rgw::d4n::BlockDirectory* blockDir = source->driver->get_block_dir();
+    block.hostsList.push_back(blockDir->cct->_conf->rgw_local_cache_address); 
+    block.cacheObj.objName = source->get_key().get_oid();
+    block.cacheObj.bucketName = source->get_bucket()->get_name();
+    std::stringstream s;
+    utime_t ut(source->get_mtime());
+    ut.gmtime(s);
+    block.cacheObj.creationTime = s.str(); 
+    block.cacheObj.dirty = false;
+
+    //populating fields needed for building directory index
+    existing_block.cacheObj.objName = block.cacheObj.objName;
+    existing_block.cacheObj.bucketName = block.cacheObj.bucketName;
+    Attrs attrs; // empty attrs for cache sets
+    std::string version = source->get_object_version();
+    if (version.empty()) {
+      version = source->get_instance();
+    }
+    ldpp_dout(dpp, 20) << __func__ << ": version stored in update method is: " << version << dendl;
+
+    if (bl.length() > 0 && last_part) { // if bl = bl_rem has data and this is the last part, write it to cache
+      std::string oid = prefix + "_" + std::to_string(ofs) + "_" + std::to_string(bl_len);
+      if (!filter->get_policy_driver()->get_cache_policy()->exist_key(oid)) { //In case of concurrent reads for the same object, the block is already cached
+        block.blockID = ofs;
+        block.size = bl.length();
+        block.version = version;
+        auto ret = filter->get_policy_driver()->get_cache_policy()->eviction(dpp, block.size, *y);
+        if (ret == 0) {
+          //Should we replace each put_async with put, to ensure data is actually written to the cache before updating the data structures and before the lock is released?
+          ret = filter->get_cache_driver()->put_async(dpp, oid, bl, bl.length(), attrs);
+          if (ret == 0) {
+            filter->get_policy_driver()->get_cache_policy()->update(dpp, oid, ofs, bl.length(), version, *y);
```

Comment:
yeah, there should probably be a version of put() that takes optional_yield so we can wait for completion here

Reply:
@cbodley : Can't we just use the blocking  `put` here, the one which is non-asynchronous. It does take in `yield` but doesn't do anything with it.

Reply:
sure. i see that `RedisDriver::put()` handles yield correctly. SSDDriver should eventually support it too, using logic similar to the CompletionToken-templated overload of `SSDDriver::get_async()`